### PR TITLE
Fix: Register service instance after provider config load

### LINKED ISSUE #686: 关于provider提前暴露ServiceInstance的解决方案
**What happened**:
provider服务元数据没有全部发布完成，应用就已经注册到了注册中心，导致consumer订阅时没有拿到完整的元数据。

**How to fix**:
去除`ServiceDiscoveryRegistry`中注册instance的代码，的在config_loader中的loadProviderConfig方法最后注册instance
具体步骤：
1、获取所有注册的`Registry`，过滤出`ServiceDiscoveryRegistry`，拿取所有`ServiceDiscovery`。
2、创建`ServiceInstance`
2、每个`ServiceDiscovery`注册`ServiceInstance`。

### PR DESCRIPTION
**Pull request by Patrick0308**
**What this PR does**:
Register service instance after provider config load

**Which issue(s) this PR fixes**:

Fixes #686

### CODE REVIEW

File: config/config_loader.go [hunk 1]
```diff
@@ -205,6 +207,98 @@ func loadProviderConfig() {
...
+	}
+	instance, err := createInstance(*url)
+	if err != nil {
+		panic(err)
+	}
+	p := extension.GetProtocol(constant.REGISTRY_KEY)
+	var rp registry.RegistryFactory
+	var ok bool
+	if rp, ok = p.(registry.RegistryFactory); !ok {
+		panic("dubbo registry protocol is invalid")
```
Comment (AlexStocks) [inline config/config_loader.go:227]: pls add rp type info in the panic string

File: config/config_loader.go [hunk 2]
```diff
@@ -205,6 +207,98 @@ func loadProviderConfig() {
...
+			logger.Errorf("url format error {%v}", url)
+			continue
+		}
+		// rest first
+		if url.Protocol == "rest" {
+			selectedUrl = url
+			break
+		} else {
+			selectedUrl = url
+		}
```
Comment (AlexStocks) [inline config/config_loader.go:299]:
```Go
    selectedUrl = url
    if url.Protocol == "rest" {
	break
    } 
```

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 3]
```diff
@@ -146,6 +151,20 @@ func (invoker *baseClusterInvoker) doSelectInvoker(lb cluster.LoadBalance, invoc
 	return selectedInvoker
 }
 
+func (invoker *baseClusterInvoker) Invoke(ctx context.Context, invocation protocol.Invocation) protocol.Result {
+	if invoker.interceptor != nil {
```
Comment (AlexStocks) [inline cluster/cluster_impl/base_cluster_invoker.go:155]: if invoker.interceptor == nil {
                 return nil
	}

	invoker.interceptor.BeforeInvoker(ctx, invocation)
	result := invoker.interceptor.DoInvoke(ctx, invocation)
	invoker.interceptor.AfterInvoker(ctx, invocation)

	return result

File: config/config_loader.go (hunk 1 above)
Comment (zouyx) [reply config/config_loader.go:229]: agree

File: config/config_loader.go [hunk 4]
```diff
@@ -206,6 +209,97 @@ func loadProviderConfig() {
...
+
+// selectMetadataServiceExportedURL get already be exported url
+func selectMetadataServiceExportedURL() *common.URL {
+	var selectedUrl common.URL
+	metaDataService, err := extension.GetMetadataService(GetApplicationConfig().MetadataType)
+	if err != nil {
+		logger.Warn(err)
+		return nil
+	}
+	list, err := metaDataService.GetExportedURLs(constant.ANY_VALUE, constant.ANY_VALUE, constant.ANY_VALUE, constant.ANY_VALUE)
```
Comment (zouyx) [inline config/config_loader.go:283]: why not  define a new variable with `var selectedUrl common.URL`

File: config/config_loader.go [hunk 5]
```diff
@@ -206,6 +209,97 @@ func loadProviderConfig() {
...
+	}
+	for _, urlStr := range list {
+		url, err := common.NewURL(urlStr.(string))
+		if err != nil {
+			logger.Errorf("url format error {%v}", url)
+			continue
+		}
+		selectedUrl = url
+		// rest first
+		if url.Protocol == "rest" {
```
Comment (zouyx) [inline config/config_loader.go:298]: if break this loop when find rest protocol, is it a problem or bug in multi protocol ?

File: registry/protocol/protocol.go [hunk 6]
```diff
@@ -117,6 +117,17 @@ func (proto *registryProtocol) initConfigurationListeners() {
 	proto.providerConfigurationListener = newProviderConfigurationListener(proto.overrideListeners)
 }
 
+func (proto *registryProtocol) GetRegistries() []registry.Registry {
```
Comment (zouyx) [inline registry/protocol/protocol.go:120]:
```suggestion
// nolint
func (proto *registryProtocol) GetRegistries() []registry.Registry {
```

File: registry/service_discovery_factory.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package registry
+
+// ServiceDiscoveryHolder
```
Comment (zouyx) [inline registry/service_discovery_factory.go:20]: add some comment for `ServiceDiscoveryHolder`

File: config/config_loader.go (hunk 4 above)
Comment (Patrick0308) [reply config/config_loader.go:284]: selectedUrl is above

File: config/config_loader.go (hunk 5 above)
Comment (Patrick0308) [reply config/config_loader.go:299]: dubbo in java use rest url first.

File: config/config_loader_test.go [hunk 8]
```diff
@@ -33,8 +37,13 @@ import (
 	"github.com/apache/dubbo-go/common/config"
 	"github.com/apache/dubbo-go/common/constant"
 	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
 	"github.com/apache/dubbo-go/common/proxy/proxy_factory"
 	"github.com/apache/dubbo-go/config_center"
+	"github.com/apache/dubbo-go/metadata/service"
+	"github.com/apache/dubbo-go/registry"
+	gxset "github.com/dubbogo/gost/container/set"
```
Comment (AlexStocks) [inline config/config_loader_test.go:45]: place gxset&gxpage into the second import block.

File: config/config_loader.go [hunk 9]
```diff
@@ -206,6 +210,97 @@ func loadProviderConfig() {
...
+			continue
+		}
+		err := sdr.GetServiceDiscovery().Register(instance)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+// createInstance
```
Comment (AlexStocks) [inline config/config_loader.go:245]: do not add comment for private func if you do not wanna add it.

**Review by zouyx — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Review by pantianying — APPROVED** [review]

**Pull request merged**